python/models/hierarchical_filtering_proxy_model.py

```diff
@@ -0,0 +1,468 @@
+            """
+            if not self.enabled:
+                return None
+
+            cache_key = self._gen_cache_key(index)
+            cache_value = self._cache.get(cache_key)
+            if not cache_value:
+                self._cache_misses += 1
+                return None
+
+            p_index, accepted = cache_value
+            if p_index and p_index == index:
+                # index and cached value are still valid!
+                self._cache_hits += 1
+                return accepted
+            else:
+                # row has changed so results are bad!
+                self._cache_misses += 1
+                return None
+
+        def minimize(self):
+            """
+            Minimize the size of the cache by removing any entries that are no longer valid
+            """
+            if not self.enabled:
+                return
+
+            self._cache = dict([(k, v) for k, v in self._cache.iteritems() if v[0].isValid()])
+
+        def clear(self):
+            """
+            Clear the cache
+            """
+            if not self.enabled:
+                return
+
+            self._cache = {}
+
+        def _gen_cache_key(self, index):
+            """
+            Generate the key for the specified index in the cache.
+
+            :param index:   The QModelIndex to generate a cache key for
+            :returns:       The key of the index in the cache
+            """
+            # ideally we would just use persistent model indexes but these aren't hashable
+            # in early versions of PySide :(
+            if self._use_persistent_index_keys:
+                return QtCore.QPersistentModelIndex(index)
+
+            # the cache key is a tuple of all the row indexes of the parent
+            # hierarchy for the index.  First, find the row indexes:
+            rows = []
+            parent_idx = index
+            while parent_idx.isValid():
+                rows.append(parent_idx.row())
+                parent_idx = parent_idx.parent()
+
+            # return a tuple of the reversed indexes:
+            return tuple(reversed(rows))
+
+    def __init__(self, parent=None):
+        """
+        Construction
+        :param parent:    The parent QObject to use for this instance
+        """
+        QtGui.QSortFilterProxyModel.__init__(self, parent)
+
+        self._accepted_cache = HierarchicalFilteringProxyModel._IndexAcceptedCache()
+        self._child_accepted_cache = HierarchicalFilteringProxyModel._IndexAcceptedCache()
+
+    def enable_caching(self, enable=True):
+        """
+        Allow control over enabling/disabling of the accepted cache used to accelerate
+        filtering.  Can be used for debug purposes to ensure the caching isn't the cause
+        of incorrect filtering/sorting or instability!
+
+        :param enable:    True if caching should be enabled, False if it should be disabled. 
+        """
+        self._dirty_all_accepted()
```

Comment:
A comment explaining why the entries become dirty just because the you are disabling or enabling caching would be appreciated.